server/service.go

```diff
@@ -28,24 +33,58 @@ func newService(svcInfo *serviceinfo.ServiceInfo, handler interface{}) *service
 }
 
 type services struct {
-	svcMap     map[string]*service
-	defaultSvc *service
+	svcSearchMap                       map[string]*service // key: "svcName.methodName" and "methodName", value: svcInfo
+	svcMap                             map[string]*service // key: service name, value: svcInfo
+	conflictingMethodHasFallbackSvcMap map[string]bool
+	fallbackSvc                        *service
+	hasCombineService                  bool
 }
 
 func newServices() *services {
-	return &services{svcMap: map[string]*service{}}
+	return &services{
+		svcSearchMap:                       map[string]*service{},
+		svcMap:                             map[string]*service{},
+		conflictingMethodHasFallbackSvcMap: map[string]bool{},
+	}
 }
 
-func (s *services) addService(svcInfo *serviceinfo.ServiceInfo, handler interface{}) {
+func (s *services) addService(svcInfo *serviceinfo.ServiceInfo, handler interface{}, isFallbackService bool) error {
 	svc := newService(svcInfo, handler)
-	if s.defaultSvc == nil {
-		s.defaultSvc = svc
+
+	// combine service check
+	if svcInfo.ServiceName == "CombineService" {
+		s.hasCombineService = true
+	} else if s.hasCombineService {
+		return errors.New("combine service cannot be registered with other services. please register services without using combine service or register only combine service")
 	}
-	s.svcMap[svcInfo.ServiceName] = svc
-}
 
-func (s *services) getService(svcName string) *service {
-	return s.svcMap[svcName]
+	if isFallbackService {
+		if s.fallbackSvc != nil {
+			return fmt.Errorf("multiple fallback services cannot be registered. [%s] is already registered as a fallback service", s.fallbackSvc.svcInfo.ServiceName)
+		}
+		s.fallbackSvc = svc
+	}
+	s.svcMap[svcInfo.ServiceName] = svc
+	for methodName := range svcInfo.Methods {
```

Comment:
I'd suggest a standalone function for such tiny tasks next time, which can be easily tested and make your function easier to understand.